### DISCUSSION TITLE
Reuse expression trees that have no closure parameters

### DISCUSSION THREAD

**Discussion by CameronAavik** [dotnet/csharplang#5381] [Language Ideas]
If you write the following code ([sharplab link](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQDdkCYAMWuowAWEAthADQAmIA1AD4ACWAjALABQjOABIywDoAMgEsAdgEcBAUQAeAByTJkIgPZjkAbk6dGAFh4AVFDAAUcxShXqAPIwCsN8TAo9gq1QBsAfN56WqEQBjWAgASh4AbwBfHQ5jZDNZHgBeP2SAUh4sVJSeHDDNIA==)):
```csharp
using System;
using System.Linq.Expressions;

void Test(Expression<Func<int, bool>> predicate) {}

Test(x => x % 2 == 0);
```
This at the time of writing would get lowered to the following code:
```csharp
ParameterExpression parameterExpression = Expression.Parameter(typeof(int), "x");
BinaryExpression body = Expression.Equal(Expression.Modulo(parameterExpression, Expression.Constant(2, typeof(int))), Expression.Constant(0, typeof(int)));
ParameterExpression[] array = new ParameterExpression[1];
array[0] = parameterExpression;
<<Main>$>g__Test|0_0(Expression.Lambda<Func<int, bool>>(body, array));
```
I think it would be good that if the expression has no closure parameters, that we reuse the same expression tree rather than recreating it each time. I am not sure how this could be implemented, but the biggest advantage of this is performance. Take the following benchmarks:
```csharp
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Linq.Expressions;

_ = BenchmarkRunner.Run<Benchmark>();

[MemoryDiagnoser]
public class Benchmark
{
    private static readonly Expression<Func<int, bool>> _expression = x => x % 2 == 0;

    [Benchmark]
    public Expression<Func<int, bool>> NonStaticExpression() => x => x % 2 == 0;

    [Benchmark]
    public Expression<Func<int, bool>> StaticExpression() => _expression;
}
```
For me this generates the following results:
|              Method |        Mean |     Error |    StdDev |  Gen 0 | Allocated |
|-------------------- |------------:|----------:|----------:|-------:|----------:|
| NonStaticExpression | 372.9492 ns | 1.5943 ns | 1.4913 ns | 0.0300 |     504 B |
|    StaticExpression |   0.0000 ns | 0.0000 ns | 0.0000 ns |      - |         - |

So we can see that reconstructing this simple expression adds a performance cost of about 370ns for each invocation and 500B of allocations. While this is a relatively small cost, you could possibly see that applications that make frequent use of expression trees such as those that use Entity Framework Linq to SQL will be incurring these costs on every request for every expression and so that will add up.

The reason that it takes 370ns to construct the expression is because of the very costly expression tree parameter validation logic that is run for every call to the `Expression` methods. It would be good instead if the compiler would be able to make the `NonStaticExpression` just as fast as the `StaticExpression` somehow.

**Comment by TahirAhmadov** [comment]
The solution to this and many other similar questions is, why not use a `static readonly` field like in your example? Whenever anything is going to be "constant", make it `const` if primitive and `static readonly` if not - simple rule, best results, no C# changes needed.

**Reply by CameronAavik** [reply]
I think there are a few reasons that this is a different scenario:
1. Expressions tend to have very long types since it needs to define all the argument types and the return type and so to move it to a field you have to declare the whole type whereas defining the expression as a parameter to a method means it infers the whole expression type for you
2. It moves the code away from the place it is being used, one of the big benefits of lambda expressions is that it is very concise syntax and so by moving this to a static field I have to go scroll up to the field (since fields usually appear at the top of the class) and it may be located far away from where it is actually used
3. It means you are unable to use anonymous types since you have to declare the full type of the expression in the static field. One can work around this by using a concrete type rather than an anonymous type but that just means having to write even more code.
4. Lambda expressions which convert to `Func<T>` and have no closure parameters will reuse the same generated closure. So I think if `Func<T>` has it, then it makes sense for `Expression<Func<T>>` to have it too.

**Reply by TahirAhmadov** [reply]
Good points. Did you look at #832 ? Local static fields would solve most of the problems with inferring delegate types, allowing anonymous types, and obviously keeping the code close to its intended usage site.
